V2.0/doc/checkpassword.3
'\"
'\"	@(#)checkpassword.3	1.11 10/15/98 (cc.utexas.edu)
'\"
.TH checkpassword 3
.SH NAME
checkpassword, configure_checkpassword, init_checkpassword checkpassword_verify \- 
password sanity checking
.sp
password_history_update \- update password history database
.sp
init_debug, set_debug_tag \- password check library tracing
.SH SYNOPSIS
.br
#include\ "checkpassword.h"
.sp
cc\ ...\ -I@NPASSWD-HOME@ -L@NPASSWD-HOME@ \ -lcheckpassword\ ...\ -lother-libs
.sp
char\ *checkpassword(char\ *password,\ struct\ passwd\ *user);
.sp
char\ *checkpassword_configure(char\ *command,\ int\ flagerror);
.sp
char\ *checkpassword_verify();
.sp
void\ init_checkpassword();
.sp
int\ password_history_update(char\ *user,\ char\ *cryptpass,\ time_t\ epoch);
.sp
void\ init_debug(int\ level);
.sp
void\ set_debug_tag(char\ *tag);
.SH DESCRIPTION
.PP
The password check library is designed for applications
which changes passwords.  This password check engine is at the core of 
.IR npasswd .
.PP
The \fBcheckpassword()\fP routine
determines whether a password is suitable for use.
The
.I password
argument is the plain-text candidate password,
.I user
is a pointer to a passwd structure from
.IR getpwent (3),
.IR getpwnam (3),
or
.IR getpwuid (3).
This routine returns an explanatory message if the password is not
acceptable, and NULL otherwise.
.PP
The \fBcheckpassword_configure()\fP routine
sets the password check parameters.
The argument
.I command
is a string containing a configuration directive.
If the
.I flagerror
argument is non-zero, unrecognized directives will be considered errors.
.PP
This routine returns NULL if
.I command
was recognized and accepted, and an error message otherwise.
.PP
The configuration directives are a subset of those for
.BR npasswd .
.PP
The \fBinit_checkpassword()\fP routine
performs initialization for the password check routines.
It is implicitly run on the first invocation of 
.BR checkpassword() .
Each call resets resets the password check parameters to defaults.
.PP
The \fBcheckpassword_verify()\fP routine sanity checks the configuration
of the password check routines. 
It returns NULL if the configuration is correct, or else an error message.
This routine should be called after the last
use of \fBinit_checkpassword()\fP or \fBcheckpassword_configure()\fP.
.PP
The \fBpassword_history_update()\fP routine inserts a password into the
history database.  The 
.I user
argument is the user name,
.I crypt-pass
is the encrypted password to enter,
and 
.I epoch
is the time stamp to put on the password.  If this argument is 0,
the current time is used.
.PP
The use of this routine requires that the history database exists
and the user has privileges to write to it.  This usually means an
effective uid of root.
.PP
This routine returns 0 if there is no history database, 1 upon success, and
-1 for failure.
.PP
If Digital UNIX enhanced security mode is active,
.B password_history_update()
is a no-op.
.PP
See manual page for
.IR history_admin (1)
about managing the password history database.
.SH INTERNALS
.PP
.B Checkpassword
subjects password candidates to a series of guessability
tests. The password is accepted only if it passes all of them.
.PP
The tests which are used and the order of their application can be
customized by configuration directives.
Some the following tests are mandatory, others optional:
.IP "History (optional)"
Password history discourages too frequent password reuse.
Password candidates are compared to the passwords in the history record
for the user and rejected if found. 
.sp
The password history database location and storage method can be
specified in the configuration file.
.PP
If Digital UNIX (OSF/1) enhanced security mode is active,
that history mechanism will be used.
.PP
.IP "Lexical (mandatory)"
These tests include:
.RS
.IP \(bu
Enforcing a minimum length of six characters.
.IP \(bu
Checking for non-printable or forbidden characters. 
Default is to allow non-printable and whitespace characters except
common tty control characters.
.IP \(bu
Denying excessive adjacent repeated characters.
.IP \(bu
Encouraging a diversity of character classes (mixed case, numbers, punctuation).
.IP \(bu
Looking for easily guessed patterns (U.S. Social Security, telephone numbers).
.IP \(bu
Passwords that do not have upper and lower case alphabetics may be rejected,
depending on the check parameters.
.RE
.PP
.IP "Local (optional)"
The password is checked against various information.
host name.
.IP "Passwd (mandatory)"
The password is compared to a number of permutations of the
information in the
.I user
argument.
.IP "Dictionary (mandatory)"
The password is subjected to a series of tests from the \fICrack\fP
password guessing program.  
The candidate is rejected if it can be derived from any word in the 
.B npasswd
password check dictionaries.
.PP
The password check library has extensive internal tracing.  The following
routines configure that facility.
.PP
The \fBset_debug_tag()\fP routine sets the trace message identifier.
.PP
The \fBinit_debug()\fP routine sets the trace level.  Refer to
.I checkpassword.h
for trace level defines.
.SH CONFIGURATION
.PP
The configuration directives are:
.TP
.BI AlphaOnly " yes | no"
Allow alpha-only passwords.  The default is to 
.B disallow
alpha-only passwords.
.TP
.BI CharClasses " N"
Require at least
.I N
character classes in passwords.
The character classes are:
.RS
.IP \(bu
Upper case alpha.
.IP \(bu
Lower case alpha.
.IP \(bu
Digits.
.IP \(bu
Whitespace.
.IP \(bu
Punctuation.
.IP \(bu
Control characters.
.RE
The higher the class requirement, the more diverse mixture of characters
required.
.TP
.BI Debug " N"
Sets internal debug level to \fIN\fP, which must be a digit [0-9].  Refer
to the sources for debug levels.
.TP
.BI Dictionaries " path ... path"
Adds directories to the dictionary search list.  Each directory in the path
is scanned for hashed dictionary files.
Multiple directories can be specified in one directive, and
multiple directives can be given.
The hash files are built with 
.BI makedict (1).
.TP
.BI DisallowedChars " [+]string"
Sets the list of characters (usually non-printable) not allowed.
C backslash sequences (e.g. \\b or \\007) may be used.
.I String
may be enclosed in double quotes.  If the initial character of
.I string
is a plus-sign, then the it is
.B appended
to the forbidden character list.
.sp
The default disallowed character list contains typical UNIX
terminal special characters.
.TP
.BI History " args"
Configure the password history mechanism.
.RS
.TP
.BI age " N"
When a password history record is fetched or updated, any passwords older than
.I N
days are discarded. The depth limit (see below) is also enforced. 
The default age is 180 days.
.TP
.BI depth " N"
When a password history record is fetched or updated, all but the
.I N
most recent 
.B old
passwords are discarded.  The age limit (see above) is also enforced.
The default is to retain the last 2 passwords.
.TP
.BI dbm " path"
Use
.I path
as the history database in DBM format. 
If
.I path
is specified as
.BR "@" ,
then the default database is used.
.TP
.BI file " path"
Use
.I path
as the history database in "flat file" format.
If
.I path
is specified as
.BR "@" ,
the default database is used.
.TP
.B none
Disable password history.
.PP
The preferred history database method is DBM, and is the
default if the NDBM library is available.
.PP
The @ syntax is useful to override the default method but use the default
database.
.PP
The default history database is 
.BR @NPASSWD-HIST@ .
.RE
.TP
.BI LengthWarn " yes | no"
Warn if the password is longer than the effective length.  This does not
affect the security of the password, but such a warning may be desirable
from a human-interface perspective.  The default is to not issue a warning.
.TP 
.BI MaxPassword " N"
Sets the maximum effective password length to \fIN\fP characters.
The default is platform dependent, but is usually 8.
.TP
.BI MaxRepeat " N"
Set allowed adjacent repeated character count to \fIN\fP.
The default is 3.
.TP
.BI MinPassword " N"
Sets the minimum acceptable length for a password to \fIN\fP characters.
The default minimum length is 6 characters.
.TP
.BI PasswordChecks " proc proc ... proc"
Selects the order of the password checks.  Use the check names listed above.
The default check order is \fBlexical\ passwd\ local\ history\ dictionary\fB.
.TP
.BI PrintableOnly " yes | no"
Limit passwords to having only printable characters.  The
.IR isprint (3)
routine is used to determine this.
The default is to allow all non-printable characters except the disallowed
characters.
Use of control characters may cause problems on some systems.
.TP
.BI SingleCase " yes | no"
Allow single-case (all lower case or all upper case) passwords.
The default is to allow single-case passwords.
.TP
.BI WhiteSpace " yes | no"
Allow whitespace characters in passwords.  The
.IR isspace (3)
routine is used to determine this.
The default is to allow whitespace characters.
.PP
.SH BUGS
.PP
The 
.B npasswd
history  database  is  protected  from general access.  Any application
using this library will need to run with privileges in order to perform
the password history check.  On Digital UNIX systems, membership in group
.I auth
is needed to read the authorization files.
.SH SEE ALSO
npasswd(1),
checkpassword(1)
.SH AUTHOR
Clyde Hoover
.br
Academic Computing Services and Instructional Technology Services
.br
The University of Texas at Austin
.br
c.hoover@cc.utexas.edu
.br
\(co 1998, The University of Texas at Austin. All rights reserved. 
'\"
'\" End checkpassword.3


v2.1/NEWS
Latest npasswd news	$Date: 1999/02/01 00:00:00 $

Patch 05	2/1/99	A passel of little bug fixes
	* Added copyright messages to more modules.
	* A bit of code format cleanup.
	* Do not enable nsswitch support unless the "nsswitch.h" include
	  file is present.  This means that nsswitch.conf will ONLY be
	  supported on Solaris 2.
	  IRIX 6.5 and Redhat Linux 4.2 have /etc/nsswitch.conf but totally
	  different API for support.

	* Properly punt the '-r' option on SunOS 5.
	* Fixed minor Makefile things throughout the distribution.
	* Added the option "-XA" to always prompt for the old password.
	* Fixed lexical password check to properly recoginize telephone numbers.
	* Enhanced debug mode in checkpassword program.
	* Use an internal pager to present help files.
	* Reorganize terminal handling code.
	* Fix -XI option to work right when stdin is a pipe
	* Fix Configure password file list input loop
	* Re-activate some Cracklib rules to be more anal about password
	  checking.

Patch 04	10/15/98	Lots of bug fixes, some feature enhacements
	* Substantial rework of history_admin:
		* Add 'merge' function (see manual page)
		* Add '-i' command line option 
		* Allow '-m' with default database
		* Fix some more annoying bugs
	* Modifications to the password history test suite to use
	  and verify the above changes
	* Clarify some issues with password history depth
		* History depth now means "how many *old* passwords matter"
		* Always retain the more recent password, even if too old
	* Allow "@" to mean the default database name in history
	  configuration directives (and the history_admin '-f' option)
	* Fix the "passwd.MaxRepeat" config directive
	* Fix the "ShadowTolerance" config directive
	* Fix the "PasswdTolerance" config directive
	* Fix the "PasswdMatchTries" config directive
	* Add build-time option "SHORT_FILE_WARN" to control behavior
	  when passwd/shadow temp files come up short
	* Support the '-r' option under SunOS 5
	* Allow change of local passwd even if NIS client (unless SunOS 4)
	* More run-time trace messages in various places

Patch 03	9/24/98
	* Fixes the "Cannot talk to NIS server" problem under Solaris 2. 

Patch 02	9/22/98
	* Fixes the build problems on Digital UNIX 4.0d.

Patch 01	9/18/98
	* Fixes build problems when NIS support is not enabled.
	* Fixes a problem with the configuration directive "PrintOnly", which
	  was renamed "PrintableOnly" during the last development cycle but
	  the change did not get propigated everywhere.

Platform status:

AIX 4.1		Builds, installs, changes local passwords
SunOS 4.1.4	Builds, installs, changes local and NIS passwords 
SunOS 5.5	Builds, installs, changes local and NIS passwords 
IRIX 6.1	Builds, installs, changes local passwords 
Digital UNIX 4.0	Builds, installs



V2.0/src/PasswordCheck/checkpassword.c
/*
 * Copyright 1998, The University of Texas at Austin ("U. T. Austin").
 * All rights reserved.
 *
 * By using this software the USER indicates that he or she has read,
 * understood and will comply with the following:
 *
 * U. T. Austin hereby grants USER permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee, provided that:
 *
 * 1. the above copyright notice appears in all copies of the software
 *    and its documentation, or portions thereof, and 
 * 2. a full copy of this notice is included with the software and its
 *    documentation, or portions thereof, and 
 * 3. neither the software nor its documentation, nor portions thereof,
 *    is sold for profit. Any commercial sale or license of this software,
 *    copies of the software, its associated documentation and/or
 *    modifications of either is strictly prohibited without the prior
 *    consent of U. T. Austin. 
 * 
 * Title to copyright to this software and its associated documentation
 * shall at all times remain with U. T. Austin. No right is granted to
 * use in advertising, publicity or otherwise any trademark, service
 * mark, or the name of U. T. Austin.
 * 
 * This software and any associated documentation are provided "as is,"
 * and U. T. AUSTIN MAKES NO REPRESENTATIONS OR WARRANTIES, EXPRESSED OR
 * IMPLIED, INCLUDING THOSE OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR
 * PURPOSE, OR THAT USE OF THE SOFTWARE, MODIFICATIONS, OR ASSOCIATED
 * DOCUMENTATION WILL NOT INFRINGE ANY PATENTS, COPYRIGHTS, TRADEMARKS OR
 * OTHER INTELLECTUAL PROPERTY RIGHTS OF A THIRD PARTY. U. T. Austin, The
 * University of Texas System, its Regents, officers, and employees shall
 * not be liable under any circumstances for any direct, indirect, special,
 * incidental, or consequential damages with respect to any claim by USER
 * or any third party on account of or arising from the use, or inability
 * to use, this software or its associated documentation, even if U. T.
 * Austin has been advised of the possibility of those damages.
 * 
 * Submit commercialization requests to: Office of the Executive Vice
 * President and Provost, U. T. Austin, 201 Main Bldg., Austin, Texas,
 * 78712, ATTN: Technology Licensing Specialist.
 */
/*
 * checkpassword.c
 *	Standalone program to use the npasswd password checker
 *
 * Compilation
 *	cc -o checkpassword checkpassword.c libpwcheck.a ../common/libcomobj.a
 */

#ifndef lint
static char sccsid[] = "@(#)checkpassword.c	1.25 11/13/98 (cc.utexas.edu) /usr/share/src/private/ut/share/bin/passwd/V2.0/src/PasswordCheck/SCCS/s.checkpassword.c";
#endif

#include <stdio.h>
#include <pwd.h>
#include "defines.h"
#include "compatibility.h"
#include "checkpassword.h"
#include "options.h"
#include "common.h"

/*
 * This program can use the npasswd configuration file.
 * The npasswd configuration file is defined in ../options.h.
 * Redefine CONFIG_FILE in the Makefile to change the default.
 */

extern char	*version,	/* Version string */
		*patchlevel,	/* Patch level string */
		*build_info;	/* Build info string */

/*
 * Program debug level table
 */
debug_table pwcheck_debug[] = {
	{ "none",	DB_NONE,	"Debugging off" },
	{ "verbose",	DB_VERBOSE,	"Verbose" },
	{ "config",	DB_CONFIG,	"Configuration processing" },
	{ "pwcheck",	DB_PWCHECK,	"Password checking" },
	{ "detail",	DB_DETAIL,	"Detailed debugging" },
	{ "all",	DB_ALL,		"All debugging" },
	{ 0 },
};

char *Usage = "Usage: %s\n\
	[-c config-file] Use alternate configuration file\n\
	[-o]		Check one password and exit\n\
	[-u user-id]	Alternate name or uid for check\n\
	[-Xh]		Print usage message\n\
	[-XDn]		Set debug level to N\n\
	[-XV]		Print version information\n\
";

/*
 * main
 *	Main program
 *
 * Exit codes
 *	0 if password is acceptable (-o only)
 *	1 if password is not acceptable (-o only)
 *	1 on error
 */
main(argc, argv, envp)
	int	argc;
	char	**argv;
	char	**envp;
{
	Uid_t	uid = getuid();		/* Invoker's uid */
	int	opt,			/* Argument parser */
		interactive = 0;	/* In interactive mode? */
	char	*config = CONFIG_FILE;	/* Configuration file */
	FILE	*pf;			/* Config file I/O */
	char	*rc;			/* Temp */
	int	oneshot = 0;		/* Check only one password switch */
	int	cfline = 0;		/* Configuration file line number */
	struct passwd *pwinfo;		/* Passwd info to check with */
	extern char	*optarg;	/* From getopt() */

	set_debug_tag("checkpassword");
	set_die_callback((void *)0);

	if ((pwinfo = getpwuid(uid)) == 0)
		die("Cannot get password info for you\n");

	/*
	 * Process command line arguments
	 */
	while ((opt = getopt(argc, argv, "c:ou:X:")) != EOF) {
		switch (opt) {
		case 'c':	/* -c config-file */
			(void) setuid(getuid());	/* Release privs */
			(void) setgid(getgid());	/* Release privs */
			config = optarg;
			if (access(config, 1) < 0)
				die("Cannot read configuration file %s\n", config);
			break;
		case 'o':	/* -o [check one password & quit] */
			oneshot++;
			break;
		case 'u':	/* -u [user id] */
#if	(CDEBUG < CDEBUG_DBX)
			if (uid)
				die("Option -u reserved for root.\n");
#endif
			if (isdigit(*optarg)) {
				Uid_t	xuid = decode_int(optarg);

				if ((pwinfo = getpwuid(xuid)) == 0) {
					die("Cannot get password for uid %d\n",
						xuid);
				}
			} else {
				pwinfo = getpwnam(optarg);
				if (pwinfo == 0)
					die("No such user '%s'\n", optarg);
			}
			break;
		case 'X':	/* -X [Extended options] */
			switch (*optarg) {
			case 'D':		/* Set debug level */
				set_debug(++optarg, pwcheck_debug);
				break;
			case 'h':		/* Help message */
				{
				debug_table *dt = pwcheck_debug;

				printf(Usage, argv[0]);
				printf("Debug levels (-XD...):\n");
				for (; dt->name; dt++)
					printf("\t%s\t%s\n", dt->name, dt->help);
				return(0);
				}
			case 'V':		/* Print version info */
				printf("Version: %s\nPatch level: %s\n%s\n",
					version, patchlevel, build_info);
				return(0);
			}
		}
	}
	checkpassword_init();
	/*
	 * Read the configuration file
	 */
	if (pf = fopen(config, "r")) {
		char	ibuf[BUFSIZ];

		while (fgets(ibuf, BUFSIZ, pf)) {
			char	*error;

			chop_nl(ibuf);
			cfline++;
			if (*ibuf == 0 || *ibuf == '#')
				continue;
			if (instringcase(ibuf, "passwd.") == 0)
				continue;

			/*
			 * Feed config line to checklib configure.
			 * Error flag is 0 so that unknown stuff is ignored
			 */
			if (error = checkpassword_configure(ibuf, 0))
				die("File '%s;; line %d; %s\n",
					config, cfline, error);
		}
		(void) fclose(pf);
	}
	if (get_debug() >= DB_DETAIL)
		checkpassword_probe("checkpassword.");

	if (rc = checkpassword_verify())
		die(rc);

	interactive = isatty(fileno(stdin));
	for (;;) {
		char	ibuf[BUFSIZ];		/* Input buffer */
		char	*nl;	/* Newline postition */

		if (interactive) {
			printf("\nPassword to check: ");
			fflush(stdout);
		}
		if (fgets(ibuf, sizeof(ibuf), stdin) == NULL) {
			if (interactive) printf("\n");
			break;
		}
		chop_nl(ibuf);
		if (ibuf[0] == 0)
			continue;
		if (rc = checkpassword(ibuf, pwinfo))
			printf("Password bad: %s.\n", rc);
		else
			printf("Password ok.\n");
		if (oneshot)
			break;
	}
	return(rc != 0);
}
/* End checkpassword.c */
